// @ledgerhq/hw-transport-http/lib/WebSocketTransport.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const hw_transport_1 = __importDefault(require("@ledgerhq/hw-transport"));
const errors_1 = require("@ledgerhq/errors");
const logs_1 = require("@ledgerhq/logs");
const WebSocket = global.WebSocket || require("ws");
/**
 * WebSocket transport implementation
 */
class WebSocketTransport extends hw_transport_1.default {
    static open(url) {
        return __awaiter(this, void 0, void 0, function* () {
            const exchangeMethods = yield new Promise((resolve, reject) => {
                try {
                    const socket = new WebSocket(url);
                    const exchangeMethods = {
                        resolveExchange: (_b) => { },
                        rejectExchange: (_e) => { },
                        onDisconnect: () => { },
                        close: () => socket.close(),
                        send: msg => socket.send(msg),
                    };
                    socket.onopen = () => {
                        socket.send("open");
                    };
                    socket.onerror = e => {
                        exchangeMethods.onDisconnect();
                        reject(e);
                    };
                    socket.onclose = () => {
                        exchangeMethods.onDisconnect();
                        reject(new errors_1.TransportError("OpenFailed", "OpenFailed"));
                    };
                    socket.onmessage = e => {
                        if (typeof e.data !== "string")
                            return;
                        const data = JSON.parse(e.data);
                        switch (data.type) {
                            case "opened":
                                return resolve(exchangeMethods);
                            case "error":
                                reject(new Error(data.error));
                                return exchangeMethods.rejectExchange(new errors_1.TransportError(data.error, "WSError"));
                            case "response":
                                return exchangeMethods.resolveExchange(Buffer.from(data.data, "hex"));
                        }
                    };
                }
                catch (e) {
                    reject(e);
                }
            });
            return new WebSocketTransport(exchangeMethods);
        });
    }
    constructor(hook) {
        super();
        this.hook = hook;
        hook.onDisconnect = () => {
            this.emit("disconnect");
            this.hook.rejectExchange(new errors_1.TransportError("WebSocket disconnected", "WSDisconnect"));
        };
    }
    exchange(apdu) {
        return __awaiter(this, void 0, void 0, function* () {
            const hex = apdu.toString("hex");
            (0, logs_1.log)("apdu", "=> " + hex);
            const res = yield new Promise((resolve, reject) => {
                this.hook.rejectExchange = (e) => reject(e);
                this.hook.resolveExchange = (b) => resolve(b);
                this.hook.send(hex);
            });
            (0, logs_1.log)("apdu", "<= " + res.toString("hex"));
            return res;
        });
    }
    setScrambleKey() { }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            this.hook.close();
            return new Promise(success => {
                setTimeout(() => {
                    success(undefined);
                }, 200);
            });
        });
    }
}
_a = WebSocketTransport;
WebSocketTransport.isSupported = () => Promise.resolve(typeof WebSocket === "function");
// this transport is not discoverable
WebSocketTransport.list = () => Promise.resolve([]);
WebSocketTransport.listen = (_observer) => ({
    unsubscribe: () => { },
});
WebSocketTransport.check = (url, timeout = 5000) => __awaiter(void 0, void 0, void 0, function* () {
    return new Promise((resolve, reject) => {
        const socket = new WebSocket(url);
        let success = false;
        setTimeout(() => {
            socket.close();
        }, timeout);
        socket.onopen = () => {
            success = true;
            socket.close();
        };
        socket.onclose = () => {
            if (success)
                resolve(undefined);
            else {
                reject(new errors_1.TransportError("failed to access WebSocketTransport(" + url + ")", "WebSocketTransportNotAccessible"));
            }
        };
        socket.onerror = () => {
            reject(new errors_1.TransportError("failed to access WebSocketTransport(" + url + "): error", "WebSocketTransportNotAccessible"));
        };
    });
});
exports.default = WebSocketTransport;
