// @zondax/zemu/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WINDOW_STAX = exports.WINDOW_X = exports.WINDOW_S = exports.KEYS = exports.DEFAULT_START_OPTIONS = exports.DEFAULT_STAX_REJECT_KEYWORD = exports.DEFAULT_STAX_APPROVE_KEYWORD = exports.DEFAULT_NANO_REJECT_KEYWORD = exports.DEFAULT_NANO_APPROVE_KEYWORD = exports.DEFAULT_METHOD_TIMEOUT = exports.KILL_TIMEOUT = exports.DEFAULT_START_TIMEOUT = exports.BASE_NAME = exports.DEFAULT_HOST = exports.DEFAULT_KEY_DELAY = exports.DEFAULT_START_DELAY = exports.DEFAULT_START_TEXT = exports.DEFAULT_MODEL = exports.DEFAULT_EMU_IMG = void 0;
exports.DEFAULT_EMU_IMG = "zondax/builder-zemu@sha256:8d7b06cedf2d018b9464f4af4b7a8357c3fbb180f3ab153f8cb8f138defb22a4";
exports.DEFAULT_MODEL = "nanos";
exports.DEFAULT_START_TEXT = "Ready";
exports.DEFAULT_START_DELAY = 20000;
exports.DEFAULT_KEY_DELAY = 100;
exports.DEFAULT_HOST = "127.0.0.1";
exports.BASE_NAME = "zemu-test-";
exports.DEFAULT_START_TIMEOUT = 30000;
exports.KILL_TIMEOUT = 5000;
exports.DEFAULT_METHOD_TIMEOUT = 30000;
exports.DEFAULT_NANO_APPROVE_KEYWORD = "APPROVE";
exports.DEFAULT_NANO_REJECT_KEYWORD = "REJECT";
exports.DEFAULT_STAX_APPROVE_KEYWORD = "Hold";
exports.DEFAULT_STAX_REJECT_KEYWORD = "Cancel";
exports.DEFAULT_START_OPTIONS = {
    logging: false,
    startDelay: exports.DEFAULT_START_DELAY,
    custom: "",
    model: exports.DEFAULT_MODEL,
    sdk: "",
    startText: exports.DEFAULT_START_TEXT,
    caseSensitive: false,
    startTimeout: exports.DEFAULT_START_TIMEOUT,
    approveAction: 9 /* ButtonKind.ApproveHoldButton */,
    approveKeyword: "",
    rejectKeyword: "",
};
exports.KEYS = {
    NOT_PRESSED: 0,
    PRESSED: 1,
    LEFT: 0xff51,
    RIGHT: 0xff53,
};
exports.WINDOW_S = {
    x: 0,
    y: 0,
    width: 128,
    height: 32,
};
exports.WINDOW_X = {
    x: 0,
    y: 0,
    width: 128,
    height: 64,
};
exports.WINDOW_STAX = {
    x: 0,
    y: 0,
    width: 400,
    height: 672,
};
