// @zondax/zemu/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WINDOW_STAX = exports.WINDOW_X = exports.WINDOW_S = exports.KEYS = exports.DEFAULT_START_OPTIONS = exports.DEFAULT_STAX_REJECT_KEYWORD = exports.DEFAULT_STAX_APPROVE_KEYWORD = exports.DEFAULT_NANO_REJECT_KEYWORD = exports.DEFAULT_NANO_APPROVE_KEYWORD = exports.DEFAULT_WAIT_TIMEOUT = exports.DEFAULT_METHOD_TIMEOUT = exports.KILL_TIMEOUT = exports.DEFAULT_START_TIMEOUT = exports.BASE_NAME = exports.DEFAULT_HOST = exports.DEFAULT_KEY_DELAY = exports.DEFAULT_START_DELAY = exports.DEFAULT_STAX_START_TEXT = exports.DEFAULT_NANO_START_TEXT = exports.DEFAULT_MODEL = exports.DEFAULT_EMU_IMG = void 0;
exports.DEFAULT_EMU_IMG = "zondax/builder-zemu:speculos-6a346803f53bc31fa8e4ef210c061fe08a9fbe16";
exports.DEFAULT_MODEL = "nanos";
exports.DEFAULT_NANO_START_TEXT = "Ready";
exports.DEFAULT_STAX_START_TEXT = "This application enables";
exports.DEFAULT_START_DELAY = 20000;
exports.DEFAULT_KEY_DELAY = 250;
exports.DEFAULT_HOST = "127.0.0.1";
exports.BASE_NAME = "zemu-";
exports.DEFAULT_START_TIMEOUT = 30000;
exports.KILL_TIMEOUT = 5000;
exports.DEFAULT_METHOD_TIMEOUT = 15000;
exports.DEFAULT_WAIT_TIMEOUT = 45000;
exports.DEFAULT_NANO_APPROVE_KEYWORD = "APPROVE";
exports.DEFAULT_NANO_REJECT_KEYWORD = "REJECT";
exports.DEFAULT_STAX_APPROVE_KEYWORD = "Hold";
exports.DEFAULT_STAX_REJECT_KEYWORD = "Cancel";
exports.DEFAULT_START_OPTIONS = {
    logging: false,
    startDelay: exports.DEFAULT_START_DELAY,
    custom: "",
    model: exports.DEFAULT_MODEL,
    sdk: "",
    startText: "",
    caseSensitive: false,
    startTimeout: exports.DEFAULT_START_TIMEOUT,
    approveAction: 10 /* ButtonKind.ApproveHoldButton */,
    approveKeyword: "",
    rejectKeyword: "",
};
exports.KEYS = {
    NOT_PRESSED: 0,
    PRESSED: 1,
    LEFT: 0xff51,
    RIGHT: 0xff53,
};
exports.WINDOW_S = {
    x: 0,
    y: 0,
    width: 128,
    height: 32,
};
exports.WINDOW_X = {
    x: 0,
    y: 0,
    width: 128,
    height: 64,
};
exports.WINDOW_STAX = {
    x: 0,
    y: 0,
    width: 400,
    height: 672,
};
